### ISSUE TITLE
How is `proofConfig` meant to be serialized?

### ISSUE THREAD

**Issue by olivereanderson** [w3c/vc-di-eddsa#21]
Part 2) of [Section 3.2 Hashing](https://w3c-ccg.github.io/di-eddsa-2020/#hashing) defines `proofConfigHash` to be the hash of `proofConfig`, but it is not stated how `proofConfig` is intended to be serialized/canonicalized in such a way that `proofConfigHash` can be unambiguously interpreted/reproduced. 

It would be much appreciated if this could be clarified. 

Thank you!

**Comment by msporny** [comment]
> Part 2) of [Section 3.2 Hashing](https://w3c-ccg.github.io/di-eddsa-2020/#hashing) defines `proofConfigHash` to be the hash of `proofConfig`, but it is not stated how `proofConfig` is intended to be serialized/canonicalized in such a way that `proofConfigHash` can be unambiguously interpreted/reproduced.

Yes, you're correct, it's quite vague at the moment... and the algorithms in VC Data Integrity and this spec just don't quite line up yet. 

The algorithms in this spec are meant to be used with the algorithms in the more generalized spec here (see steps 2 and 3):

https://w3c.github.io/vc-data-integrity/#generate-proof

The problem, of course, is that implementers then need to jump back and forth between specs (which is awful). I'm currently trying to figure out a way to make this jumping back and forth easier. I've been thinking of pointing from the Data Integrity specification to the subalgorithms in each cryptosuite spec (and vice-versa)... and/or figuring out a way to transclude the data integrity steps into each cryptosuite spec (but have to write a bunch of code to do that).

The answer to your question is you run the algorithms in this order:

1. https://w3c.github.io/vc-data-integrity/#generate-proof
1.1 https://w3c-ccg.github.io/di-eddsa-2020/#transformation (on the unsecured document)
1.2 https://w3c-ccg.github.io/di-eddsa-2020/#proof-configuration
1.2.1 https://w3c-ccg.github.io/di-eddsa-2020/#transformation (on the `proofOptions` from the previous step)
1.3 https://w3c-ccg.github.io/di-eddsa-2020/#hashing (on both the `transformedDocument` and `proofOptions`)
1.x and then you run the rest of the algorithm to finish up the signature

You got caught between a reorganization of the algorithms in both specs, apologies for that.

I know that's not as clear as it needs to be, but does that help?

My hope is that we can get these algorithms cleaned up by the March 2023 timeframe.

**Comment by olivereanderson** [comment]
@msporny Thank you for your reply, 

The answer you gave to my question is understandable! 

Since I have your attention now could I ask a few more related questions here? 

Suppose I am given a Verifiable credential with a proof generated by this suite and I verify the proof according to this suite using a JSON-LD processor satisfying the the two requirements specified in the [semantic interoperability section of the VC datamodel V2 spec](https://www.w3.org/TR/vc-data-model-2.0/#semantic-interoperability) in the transformation/canonicalization step. If proof verification checks out can I then proceed with custom validations of the credential (for instance check that `Credential.ValidFrom` is more than two weeks ago) _without_ using a JSON-LD processor for these last validations? Or are hacks of the kind demonstrated in the following [github repo](https://github.com/setl/rdf-urdna/tree/master/jsonld-warnings) still applicable at this point (hence all validation needs to be done using a JSON-LD proessor)? 

Thanks once more for your time.

**Comment by msporny** [comment]
> Suppose I am given a Verifiable credential with a proof generated by this suite and I verify the proof according to this suite using a JSON-LD processor satisfying the the two requirements specified in the [semantic interoperability section of the VC datamodel V2 spec](https://www.w3.org/TR/vc-data-model-2.0/#semantic-interoperability) in the transformation/canonicalization step. If proof verification checks out can I then proceed with custom validations of the credential (for instance check that `Credential.ValidFrom` is more than two weeks ago) _without_ using a JSON-LD processor for these last validations? 

Yes, that is correct (in general). We have tried really hard to make sure you can process these objects using standard JSON tooling /after/ you've verified the data integrity proof (digital signature).

You have to make sure that the object you're working with isn't a pre-verified object, but rather a post-verified one... that is, a pre-verified object might contain bad data whereas a post-verified object will only contain data that was digitally signed and valid from a JSON-LD perspective. You might also want to run a JSON Schema against the object to make sure it fits the schema your application is expecting.

I hope the above makes sense.

> Or are hacks of the kind demonstrated in the following [github repo](https://github.com/setl/rdf-urdna/tree/master/jsonld-warnings) still applicable at this point (hence all validation needs to be done using a JSON-LD proessor)?

Yikes, that github repo is all sorts of misleading, if not flat out wrong.

> 1. You can add data with very few restrictions and it does not change the signature. See "hack1" for an example.

It is true that data that is invalid JSON-LD will be dropped when signing (and verifying). There are caveats to this, like how the latest Verifiable Credentials Data Model now has a default vocabulary so undefined terms will be signed over now (changing the signature if you try to include undefined terms). Some see that as a benefit, others see it as a drawback.

> 2. You can rename or move any property without changing the signature. See "hack2".

Misleading. The only time that's true is if you have no term mappings for the properties ... which means those properties are dropped... which means that they're not signed over.

> 3. You can replace any string value with a different string without changing the signature. See "hack2"

Again, misleading. If you don't have term mappings for a property, it's dropped and not signed over.

The three items above are effectively: "I am not using the digitally signed object to drive my application... instead, I'm using a pre-signed object."

> 4. You can submit small documents that take years to process.

Sure, if you turn `safe` mode off... don't do that unless you know what you're doing. 

This issue is called a "poison graph" and is a general problem in any graph-based data structure (not just RDF) -- See the intro here for more information on that topic: https://www.w3.org/TR/rdf-canon/#introduction

It turns out that it's not a problem with real world datasets and if someone tries to attack you with such a data structure, any library doing RDF Dataset Canonicalization should turn safe mode on by default:

https://github.com/digitalbazaar/jsonld.js/blob/main/lib/jsonld.js#L523-L531

In short, all of the issues that the individual points out above are a result of NOT using the post-verified object and using the pre-verified object instead. The guidance to not use blank nodes is bad guidance as it's not blank nodes that are the issue, but blank nodes that have zero differentiating characteristics from other blank nodes... and again, with safe mode turned on by default, you're not susceptible to that attack.

**Comment by msporny** [comment]
@olivereanderson I'm currently reworking all of the algorithms to make them a little easier to read. Here's a concrete realization of the algorithms, please let me know if you find them easier (or harder) to follow than before:

https://w3c.github.io/vc-di-eddsa/#eddsa-2022

You're supposed to start reading at "Add Proof" or "Verify Proof" in the cryptosuite, which will refer you back to the generalized algorithms.

It's all still a bit piecemeal, with the generalized algorithm being defined in the Data Integrity specification and the cryptography suite-specific algorithms being defined in the cryptographic suites. 

I know that the ideal would be to just look at ONE document and read it from top to bottom, but due to the way the generalized algorithms are defined in Data Integrity and the specific algorithms are defined in each cryptosuite... I'd have to write quite a bit of code to autogenerate the algorithmic prose in the specifications. It's doable, but not without great effort.

In any case, let me know if it's a little easier to follow now. I'll keep improving the algorithms as the weeks roll on.

**Comment by Wind4Greg** [comment]
Hi all, I've been working on creating test vectors for the specification. I put together a minimalistic JavaScript creation example here: https://gist.github.com/Wind4Greg/a601a4617fa2963da2bcce39303283e3. This also shows some of the missing pieces that @olivereanderson mentioned. Cheers

**Comment by olivereanderson** [comment]
@msporny Thank you for your explanations and for improving the specification.  

Although the specification is more readable now I still think it needs to clearly specify that the `proofConfig` should be canonicalized before being hashed as illustrated in [this line of code](https://gist.github.com/Wind4Greg/a601a4617fa2963da2bcce39303283e3#file-vcintegritytestcreate2-js-L104) in the example posted by @Wind4Greg above. As far as I can tell that step is still missing from the specification (and I could also not find it in the [VC Data Integrity specification](https://w3c.github.io/vc-data-integrity)). 

Again thank you for answering my questions and taking the time to work on this!

**Comment by Wind4Greg** [comment]
Did PR https://github.com/w3c/vc-di-eddsa/pull/31 and https://github.com/w3c/vc-di-eddsa/pull/43 resolve this issue? Can we close the issue?

Cheers Greg
